common/lib/Warewulf/DataStore/SQL/PostgreSQL.pm

```diff
@@ -0,0 +1,294 @@
+          LEFT JOIN lookup ON lookup.object_id = datastore.id
+END_OF_SQL
+    if (@query_opts) {
+        $sql_query .= ' WHERE '. join(' AND ', @query_opts);
+    }
+    $sql_query .= ' GROUP BY datastore.id';
+
+    return $sql_query;
+}
+
+
+sub
+get_lookups_build_query_impl()
+{
+    my $self = shift;
+    my ($type, $field, $paramsRef, @strings) = @_;
+    
+    my $sql_query = <<'END_OF_SQL';
+        SELECT
+            lookup.value AS value
+          FROM lookup
+          LEFT JOIN datastore ON lookup.object_id = datastore.id
+          WHERE
+            lookup.field != 'ID'
+END_OF_SQL
+    my @predicates;
+    
+    if ($type) {
+        push(@$paramsRef, $type);
+        push(@predicates, 'datastore.type = ?');
+    }
+    if ($field) {
+        push(@$paramsRef, uc($field));
+        push(@predicates, 'lookup.field = ?');
+    }
+    if (@strings) {
+        foreach my $s (@strings) {
+            push(@$paramsRef, $s);
+        }
+        push(@predicates, 'lookup.value IN (' . substr(',?' x scalar(@strings), 1) . ')');
+    }
+
+    if (@predicates) {
+        $sql_query .= ' AND '. join(' AND ', @predicates);
+    }
+    $sql_query .= ' GROUP BY lookup.value';
+
+    return $sql_query;
+}
+
+
+sub
+allocate_object_impl()
+{
+    my $self = shift;
+    my ($type) = @_;
+
+    if (!exists($self->{"STH_INSTYPE"})) {
+        $self->{"STH_INSTYPE"} = $self->{"DBH"}->prepare("INSERT INTO datastore (type) VALUES (?)");
+    }
+    $self->{"STH_INSTYPE"}->execute($type);
+    return $self->{"DBH"}->last_insert_id(undef, undef, 'datastore', 'id');
+}
+
+
+sub
+update_datastore_impl()
+{
+    my $self = shift;
+    my ($id, $serialized_data) = @_;
+    
+    if (!exists($self->{"STH_SETOBJ"})) {
+        $self->{"STH_SETOBJ"} = $self->{"DBH"}->prepare("UPDATE datastore SET serialized = ? WHERE id = ?");
+    }
+    $self->{"STH_SETOBJ"}->bind_param(1, $serialized_data, { pg_type => DBD::Pg::PG_BYTEA });
+    $self->{"STH_SETOBJ"}->bind_param(2, $id);
+    my $rc = $self->{"STH_SETOBJ"}->execute();
+    $self->{"STH_SETOBJ"}->finish();
+    return $rc;
+}
+
+
+sub
+put_chunk_impl()
+{
+    my ($self, $buffer) = @_;
+
+    if (!exists($self->{"STH_PUT"})) {
+        $self->{"STH_PUT"} = $self->{"DBH"}->prepare("INSERT INTO binstore (object_id, chunk) VALUES (?,?)");
+        $self->{"DBH"}->do("DELETE FROM binstore WHERE object_id = ?", undef, $self->{"OBJECT_ID"});
+        &dprint("SQL: INSERT INTO binstore (object_id, chunk) VALUES ($self->{OBJECT_ID},?)\n");
+    }
+    
+    $self->{"STH_PUT"}->bind_param(1, $self->{"OBJECT_ID"});
+    $self->{"STH_PUT"}->bind_param(2, $buffer, { pg_type => DBD::Pg::PG_BYTEA });
+    my $rc = $self->{"STH_PUT"}->execute();
+    $self->{"STH_PUT"}->finish();
+    if ( ! $rc ) {
+        &eprintf("put_chunk() failed with error:  %s\n", $self->{"STH_PUT"}->errstr());
+        return 0;
+    }
+    return 1;
+}
+
+
+=back
+
+=head1 SEE ALSO
+
+Warewulf::ObjectSet Warewulf::DataStore
+
+=head1 COPYRIGHT
```

Comment:
I cede all copyright stuff to the project; if accepted, my small contribution will at least be documented here. :-)